schema/spacewalk/postgres/tables/suseSaltEvent.sql

```diff
@@ -3,7 +3,8 @@
 CREATE TABLE suseSaltEvent (
   id SERIAL PRIMARY KEY,
   minion_id CHARACTER VARYING(256),
-  data TEXT NOT NULL
+  data TEXT NOT NULL,
+  queue NUMERIC NOT NULL
 );
 
 CREATE INDEX suse_salt_event_minion_id_idx
```

Comment:
I think we want to index `queue` as well now. Not sure if it's best to add it to the existing index, or create a new one, needs Postgres investigation. What we definitely want is the `popSaltEvents` query to be very fast, possibly `countSaltEvents` too. It might or might not make sense to also `ORDER BY queue` in `popSaltEvents`.

Reply:
I think you are the postgresql optimization expert around, isn't it?

Reply:
Ha! Thanks, I'll take that as a compliment :wink:

I did some tests and the following emerged:

1- we cannot add a `NOT NULL` column to a non-empty table, and we can't guarantee it will be empty. Nevertheless we can assume not a lot of events will be in there at migration time, so I would suggest to change the migration to something like:

```sql
ALTER TABLE suseSaltEvent ADD queue NUMERIC;
UPDATE suseSaltEvent SET queue = 0;
ALTER TABLE suseSaltEvent ALTER COLUMN queue SET NOT NULL;
```

2- in order for `popSaltEvents` and `countSaltEvents` to use the index, we should add `queue` to it. Specifically this does the trick:

```sql
DROP INDEX IF EXISTS suse_salt_event_minion_id_idx;

CREATE INDEX IF NOT EXISTS suse_salt_event_minion_id_idx
  ON suseSaltEvent (queue, minion_id NULLS FIRST, id);
```

Both the seed schema and the upgrade script should reflect this.

Reply:
:heavy_check_mark: